idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,987 @@
 
 __author__ = "Alexander Dowling, Douglas Allan, Andrew Lee"
 
-
 from operator import itemgetter
+from sys import stdout
 
-import pyomo.environ as pyo
-from pyomo.core.expr.visitor import identify_variables
-from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
-from pyomo.contrib.pynumero.asl import AmplInterface
-from pyomo.core.base.block import _BlockData
 import numpy as np
 from scipy.linalg import svd
 from scipy.sparse.linalg import svds, norm
 from scipy.sparse import issparse, find
 
+from pyomo.environ import (
+    Binary,
+    Block,
+    check_optimal_termination,
+    ConcreteModel,
+    Constraint,
+    Objective,
+    Param,
+    Set,
+    SolverFactory,
+    value,
+    Var,
+)
+from pyomo.core.base.block import _BlockData
+from pyomo.common.collections import ComponentSet
+from pyomo.common.config import ConfigDict, ConfigValue, document_kwargs_from_configdict
+from pyomo.util.check_units import identify_inconsistent_units
+from pyomo.contrib.incidence_analysis import IncidenceGraphInterface
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+from pyomo.contrib.pynumero.asl import AmplInterface
+from pyomo.common.deprecation import deprecation_warning
+
 from idaes.core.util.model_statistics import (
+    activated_blocks_set,
+    deactivated_blocks_set,
+    activated_equalities_set,
+    deactivated_equalities_set,
+    activated_inequalities_set,
+    deactivated_inequalities_set,
+    activated_objectives_set,
+    deactivated_objectives_set,
+    variables_in_activated_constraints_set,
+    variables_not_in_activated_constraints_set,
+    degrees_of_freedom,
     large_residuals_set,
     variables_near_bounds_set,
 )
-import idaes.core.util.scaling as iscale
+from idaes.core.util.scaling import (
+    get_jacobian,
+    extreme_jacobian_columns,
+    extreme_jacobian_rows,
+    extreme_jacobian_entries,
+    jacobian_cond,
+)
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+MAX_STR_LENGTH = 84
+TAB = " " * 4
+
+# TODO: Add suggested steps to cautions - how?
+
+CONFIG = ConfigDict()
```

Comment:
I'm a little concerned that there are too many config options, which will be hard to remember. Especially since some of these could be easily confused (e.g. one might misuse `zero_value_tolerance` thinking that it applies to residuals).

What is the minimum set of tolerances that any single method needs? It seems like just a "zero tolerance" and possibly a "small tolerance". Then large value thresholds can just default to the inverse of these tolerances (or `float("inf")` if either is exactly zero). 

What do you think about a ConfigDict like this:
```python
from pyomo.common.config import ConfigValue, ConfigDict

CONFIG = ConfigDict()
CONFIG.declare("tolerance", ConfigValue(default=1e-8, domain=float))
CONFIG.declare("caution_tolerance", ConfigValue(default=1e-4, domain=float))
CONFIG.declare("large_value_threshold_warning", ConfigValue(default=1e8, domain=float))
CONFIG.declare("large_value_threshold_caution", ConfigValue(default=1e4, domain=float))
```
Then individual methods can just use e.g. `tolerance` for whatever is relevant for the particular method. This would work better if the individual methods can accept `**kwds`, so that one can call two methods with different tolerances, without having to know exactly what the right config arg for each method is:
```python
dt.display_variable_at_or_outside_bounds(tolerance=1e-6)
dt.display_constraints_with_extreme_jacobians(tolerance=1e-10)
```

I guess the downside of this is that we can't fine-tune each individual tolerance then call `report_numerical_issues`. There is also a potential backward compatibility issue if we implement a general `tolerance` arg then later decide we need more fine-grained options.

Reply:
Maybe the way to go is to just stick with the fine-grained tolerance options, but add some way to print all the config options so they can be easily discovered.

Reply:
Firstly, the config block is automatically included in the class doc string (via the Pyomo decorator magic), so user can easily see all the config arguments. This also extends to the documentation generated by Sphinx so it is all easy to see.

Secondly, I did start with fewer tolerances, but came to found that I wanted most of them as I generally wanted different tolerances for the Jacobian versus the residuals, etc.

Reply:
> Firstly, the config block is automatically included in the class doc string (via the Pyomo decorator magic), so user can easily see all the config arguments. This also extends to the documentation generated by Sphinx so it is all easy to see.

Great, I didn't know (or forgot) that. I see now that what I'm asking for is a subset of what `help(DiagnosticsToolbox)` already does.